zebra/zebra_rib.c

```diff
@@ -4376,6 +4383,23 @@ int rib_add_multipath(afi_t afi, safi_t safi, struct prefix *p,
 		nhe.id = re->nhe_id;
 
 	n = zebra_nhe_copy(&nhe, 0);
+
+	if (re->type == ZEBRA_ROUTE_KERNEL) {
+		struct interface *ifp;
+
+		if (p->family == AF_INET6 &&
+		    IN6_IS_ADDR_LINKLOCAL(&p->u.prefix6)) {
+			XFREE(MTYPE_RE, re);
+			return -1;
+		}
+
+		ifp = if_lookup_prefix(p, re->vrf_id);
+		if (ifp) {
+			if (ifp->ifindex == ng->nexthop->ifindex)
+				re->type = ZEBRA_ROUTE_CONNECT;
```

Comment:
I would prefer that we keep things the way we have them as that this is how we think about the problem in FRR.  I'm not sure that it matters though.

Reply:
to clarify, my point a bit more.  I do not want any route that has link scope to become a connected route.  A connected route is a route that has an address associated with it on the interface in question.  That is the definition that I want to stick with.

Reply:
Right, okay. My concern was mostly that by not mirroring the exact semantics of the kernel FRR can end up with a different view of how packets will be routed than what the forwarding plane actually does. However, I don't have a concrete use case where this will lead to problems, and I guess for the original "set noprefixroute but install another replacement route for the prefix in the kernel" use case, this is fine :)

Reply:
be aware that this patch will just turn those local scope routes into kernel routes so nothing to worry about

Reply:
Ah, right, so FRR will still end up with a consistent view of the forwarding state (because it no longer skips all `RTPROT_KERNEL` routes), but only those with a matching interface prefix will be considered connected? Makes sense!